server/src/main/java/org/elasticsearch/search/aggregations/InternalMultiBucketAggregation.java

```diff
@@ -141,6 +142,28 @@ public static int countInnerBucket(Aggregation agg) {
         return size;
     }
 
+    /**
+     * Unlike {@link InternalAggregation#materializePipelines(InternalAggregation, ReduceContext)}, a multi-bucket
+     * agg needs to first materialize the buckets (and their parent pipelines) before allowing sibling pipelines
+     * to materialize
+     */
+    @Override
+    public final InternalAggregation materializePipelines(InternalAggregation reducedAggs, ReduceContext reduceContext) {
+        assert reduceContext.isFinalReduce();
+        List<B> materializedBuckets = materializeBuckets(reduceContext);
+        return super.materializePipelines(create(materializedBuckets), reduceContext);
+    }
+
+    private List<B> materializeBuckets(ReduceContext reduceContext) {
+        return getBuckets().stream().map(internalBucket -> {
```

Comment:
I'm -0 on using streams here.  IMHO, Streams are only good when they serve to enhance readability, and I'm not sure they're doing that here, relative to the equivalent loops.  I'm ok with leaving it this way if you prefer it, but it's not my first choice.  Unless of course you've already looked at the loop based version and decided it's even harder to read, in which case I guess this is the lesser evil.